src/lib.rs

```diff
@@ -197,62 +183,144 @@ where
         match &self.signer {
-            Some(key) => {
+            Some(pair) => {
                 let mut arr: [u8; 32] = Default::default();
-                arr.clone_from_slice(key.to_owned().public().as_ref());
-                Ok(Self::_get_nonce(self.url.clone(), arr))
+                arr.clone_from_slice(pair.to_owned().public().as_ref());
+                let accountid: AccountId = Decode::decode(&mut &arr.encode()[..]).unwrap();
+                if let Some(info) = self.get_account_info(&accountid) {
+                    Ok(info.nonce)
+                } else { Ok(0) }
             }
-            None => Err("Can't get nonce when no signer is set"),
+            None => Err("Can't get )nonce when no signer is set"),
         }
     }
 
-    pub fn get_account_data(&self, address: &AccountId32) -> Option<AccountData<u128>> {
+    pub fn get_account_info(&self, address: &AccountId) -> Option<AccountInfo> {
         let id: &[u8; 32] = address.as_ref();
-        let result_str = self
-            .get_storage("Balances", "Account", Some(id.to_owned().encode()))
-            .unwrap();
+        let storagekey: sp_core::storage::StorageKey = self.metadata
+            .module("System").unwrap()
+            .storage("Account").unwrap()
+            .get_map::<AccountId,AccountInfo>().unwrap()
+            .key(address.clone());
+        info!("storagekey {:?}",storagekey);
+        info!("storage key is: 0x{}", hex::encode(storagekey.0.clone()));
+        self.get_storage_by_key_hash(storagekey.0)
+    }
+
+    pub fn get_account_data(&self, address: &AccountId) -> Option<AccountData> {
+        if let Some(info) = self.get_account_info(address) {
+            Some(info.data)
+        } else { None }
+    }
+
+    pub fn get_finalized_head(&self) -> WsResult<String> {
+        Self::_get_request(self.url.clone(), json_req::chain_get_finalized_head().to_string())
+    }
+
+    pub fn get_header(&self, hash: Option<Hash>) -> WsResult<String> {
+        Self::_get_request(self.url.clone(), json_req::chain_get_header(hash).to_string())
+    }
 
-        hexstr_to_account_data(result_str).ok()
+    pub fn get_block(&self, hash: Option<Hash>) -> WsResult<String> {
+        Self::_get_request(self.url.clone(), json_req::chain_get_block(hash).to_string())
     }
 
     pub fn get_request(&self, jsonreq: String) -> WsResult<String> {
         Self::_get_request(self.url.clone(), jsonreq)
     }
 
-    pub fn get_storage(
+    pub fn get_storage_value<V:Decode+Clone>(
         &self,
-        storage_prefix: &str,
-        storage_key_name: &str,
-        param: Option<Vec<u8>>,
-    ) -> WsResult<String> {
-        Self::_get_storage(self.url.clone(), storage_prefix, storage_key_name, param)
+        storage_prefix: &'static str,
+        storage_key_name: &'static str,
+    ) -> Option<V> {
+        let storagekey: sp_core::storage::StorageKey = self.metadata
+                .module(storage_prefix).unwrap()
```

Comment:
Although this is not critial code, I am a little bit concerned about these `unwrap()`s. 

As they all return `MetadataError`,  would recommend to make the function return: `Result<V, MetadataError>` and use `module()?`